### PR TITLE
[GStreamer][WebRTC] Emit 'track' event before resolving setRemoteDescription's promise

### PR DESCRIPTION
**Pull request by cadubentzen**
#### 584f20310cd3710ada27d5b2498c3c4d90df5f09
<pre>
[GStreamer][WebRTC] Emit &apos;track&apos; event before resolving setRemoteDescription&apos;s promise
<a href="https://bugs.webkit.org/show_bug.cgi?id=275685">https://bugs.webkit.org/show_bug.cgi?id=275685</a>

Reviewed by Philippe Normand.

The track event should be emitted after sucessfully applying a remote
description and before the promise from setRemoteDescription is resolved.
However, up until now the GstWebRTC backend emitted the event based on
the &apos;pad-added&apos; signal. The timing of the signal doesn&apos;t match what the
spec expects.

Fix the issue by querying the transceiver states after applying a remote
description sucessfully. By providing the transceiver states to
PeerConnectionBackend::setRemoteDescriptionSucceeded(), it&apos;s able to
emit the track events correctly acccording to the spec.

A test is added to check that the track event is emitted before
setRemoteDescription resolves, which would fail before this patch in
the GstWebRTC backend.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-inspect-answer-expected.txt:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-media-setup-single-dialog-expected.txt:
* LayoutTests/platform/glib/fast/mediastream/RTCPeerConnection-setRemoteDescription-offer-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-setLocalDescription-parameterless.https-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webrtc/protocol/split.https-expected.txt: Added.
* LayoutTests/webrtc/setRemoteDescription-track-expected.txt: Added.
* LayoutTests/webrtc/setRemoteDescription-track.html: Added.
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::setRemoteDescriptionSucceeded):
(WebCore::PeerConnectionBackend::setRemoteDescriptionFailed):
(WebCore::PeerConnectionBackend::stop):
(WebCore::PeerConnectionBackend::dispatchTrackEvent): Deleted.
(WebCore::PeerConnectionBackend::addPendingTrackEvent): Deleted.
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::initializePipeline):
(WebCore::GStreamerMediaEndpointTransceiverState::isolatedCopy):
(WebCore::getMediaStreamIdsFromSDPMedia):
(WebCore::isRecvDirection):
(WebCore::toGStreamerMediaEndpointTransceiverState):
(WebCore::transceiverStatesFromWebRTCBin):
(WebCore::GStreamerMediaEndpoint::doSetLocalDescription):
(WebCore::GStreamerMediaEndpoint::setTransceiverCodecPreferences):
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::setDescription):
(WebCore::GStreamerMediaEndpoint::connectIncomingTrack):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::addPendingTrackEvent): Deleted.
(WebCore::GStreamerPeerConnectionBackend::dispatchPendingTrackEvents): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::dataFlowStarted):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
(WebCore::MediaStreamTrackPrivateObserver::dataFlowStarted):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(webkitMediaStreamSrcCharacteristicsChanged):

Canonical link: <a href="https://commits.webkit.org/281892@main">https://commits.webkit.org/281892@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/80af5716fa7f6913dfadbb26a68878851059c479

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61378 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13962 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65330 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12201 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49584 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8281 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53164 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30427 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10839 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56339 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67061 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10466 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56957 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53129 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57172 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13676 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4396 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36542 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37625 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38719 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37369 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp [hunk 1]
```diff
@@ -170,6 +170,11 @@ bool GStreamerMediaEndpoint::initializePipeline()
     }), this);
 
 #ifndef GST_DISABLE_GST_DEBUG
+    g_signal_connect_swapped(m_webrtcBin.get(), "on-new-transceiver", G_CALLBACK(+[](GStreamerMediaEndpoint* endPoint, GstWebRTCRTPTransceiver* transceiver) {
+        GST_DEBUG_OBJECT(endPoint->m_webrtcBin.get(), "New transceiver: %" GST_PTR_FORMAT, transceiver);
+    }), this);
+
```
Comment (philn) [inline Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:176]: Nit: Can you remove this extra empty line?

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp [hunk 2]
```diff
@@ -371,6 +376,125 @@ static std::optional<PeerConnectionBackend::DescriptionStates> descriptionsFromW
...
+    }
+#endif
+
+    Vector<String> streamIds;
+    if (remoteDescription && remoteDescription->sdp && mLineIndex != std::numeric_limits<guint>::max()) {
+        const GstSDPMedia* media = gst_sdp_message_get_media(remoteDescription->sdp, mLineIndex);
+
+        if (!mid) {
+            const auto midFromSDP = gst_sdp_media_get_attribute_val(media, "mid");
+            mid.outPtr() = g_strdup(midFromSDP);
```
Comment (philn) [inline Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp]:
```suggestion
            mid.reset(g_strdup(midFromSDP));
```

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp [hunk 3]
```diff
@@ -1595,16 +1768,24 @@ void GStreamerMediaEndpoint::collectTransceivers()
...
         }
-
-        m_peerConnectionBackend.newRemoteTransceiver(WTF::makeUnique<GStreamerRtpTransceiverBackend>(WTFMove(current)), m_mediaForMid.get(String::fromUTF8(mid.get())), trackIdFromSDPMedia(*media));
+        auto transceiverBackend = WTF::makeUnique<GStreamerRtpTransceiverBackend>(WTFMove(current));
+        // As of GStreamer 1.24.4, webrtcbin's transceivers only have the "mid" property set after the signaling state
+        // is STABLE. However, we need to use the transceiver mid property when signaling is still HAVE_REMOTE_OFFER.
+        // For that, we set the mid from the SDP directly as an alternative for using it in this early stage.
+        // FIXME: make this version-dependent when it's fixed in a released version of GStreamer.
+        if (!mid && signalingState == GST_WEBRTC_SIGNALING_STATE_HAVE_REMOTE_OFFER) {
+            mid.outPtr() = g_strdup(gst_sdp_media_get_attribute_val(media, "mid"));
```
Comment (philn) [inline Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp]:
```suggestion
            mid.reset(g_strdup(gst_sdp_media_get_attribute_val(media, "mid")));
```

**Comment by philn** [comment]
With this fixup the CI should be happier: https://github.com/philn/WebKit/commit/85523bf394a222cce2035b355849cf0d099127aa

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp (hunk 2 above)
Comment (cadubentzen) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp]: `GUniqueOutPtr::reset()` is private and takes no args. https://github.com/WebKit/WebKit/pull/30112

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp (hunk 3 above)
Comment (cadubentzen) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp]: not needed anymore if we depend on the fix in webrtcbin

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp (hunk 2 above)
Comment (cadubentzen) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp]: not needed anymore if we depend on the fix in webrtcbin.

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp (hunk 1 above)
Comment (cadubentzen) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:176]: done

**Comment by philn** [comment]
I guess EWS fails because the current SDK doesn't include the GStreamer patch, right?

**Comment by philn** [comment]
The style lint errors should be fixed too please ;)

**Comment by cadubentzen** [comment]
> I guess EWS fails because the current SDK doesn't include the GStreamer patch, right?

exactly

**Review by philn — APPROVED** [review]

**Pull request merged**